lib/yeti_exception/helpers.rb

```diff
@@ -0,0 +1,113 @@
+module YetiException
+
+  # Mixin for constructing and raising YetiException::Error or its subclasses.
+  # Two convenience methods are provided, both on the class and on class
+  # instances:
+  #
+  # - raise_exception: Infers the class in which the exception is raised so that
+  #   it can be used in constructing the exception.
+  #
+  # - wrap_exception: Creates a new YetiException exception from an arbitrary
+  #   exception, maintaining the orignal's backtrace.  This is useful when
+  #   dealing with exceptions from third-party code, so that the extra features
+  #   of YetiException may be used without losing the location of the original
+  #   error.
+  #
+  # Example usage:
+  #
+  #     class MyClass
+  #       include YetiException::Helpers
+  #
+  #       def foo(x)
+  #         if x.nil?
+  #           raise_exception(YetiException::Error,
+  #                           {
+  #                             msg: 'x is nil',
+  #                             something: 'interesting'
+  #                           })
+  #         end
+  #       end
+  #
+  #       def bar(y)
+  #         ThirdParty.Bar(y)
+  #       rescue => ex
+  #         wrap_exception(ex,
+  #                        YetiException::Error,
+  #                        {
+  #                          y_value: y
+  #                        })
+  #       end
+  #     end
+
+  module Helpers
+
+    # Raise the specified exception, inferring the calling class automatically.
+    #
+    # @param exception_klass [Class]
+    # @param details [Hash]
+    # @param args [Object ...]
+    # @raise [YetiException::Error]
+    def raise_exception(exception_klass, details, *args)
+      raise exception_klass.new(self.class, details, *args)
+    end
+
+    # Wrap an arbitrary exception in a new enhanced exception and reraise,
+    # preserving the original backtrace and merging the original message into
+    # the details.
+    #
+    # @param original_exception [Exception]
+    # @param new_exception_klass [Class]
+    # @param details [Hash]
+    # @param args [Object ...]
+    # @raise [YetiException::Error]
+    def wrap_exception(original_exception, new_exception_klass, details, *args)
```

Comment:
I found this to be a confusing feature when we used it in the tardis APIs. A number of errors would claim to come from the API class, but were actually occurring in the helper procs or client gems.

Reply:
That's why the original exception's error message and backtrace are preserved.

Reply:
Does [`Exception#cause`](https://ruby-doc.org/core-2.6/Exception.html#method-i-cause) help us at all? Can we use `e.cause.backtrace` instead of setting the backtrace here? I wonder if `.cause` shows up in logs.

Reply:
With error handling at the edges of the application and error handling at various depths to merge higher level details into the errors being thrown from deeper levels in the stack, it's less intuitive to me to see: `'MidLevelError: deep error message mid level details at deep_file.rb'` over `'DeepError: deep error message mid level details at deep_file.rb'` - the former is not a 'mid level' exception. How do we distinguish between a real mid level error from a re-raise mid level error? Could we merge the details into the original error / raise a new exception of the original exception class?

Reply:
I'll have to look into `#cause`.  I'm concerned that you'd need to be aware of it and poke deeper.  For example, if the `YetiException::Error` was raised on its own, meaning there was no previous exception, is `#cause` nil?  And so you need to include it conditionally?  I don't see any reference to `cause` in YetiLogger, so it must not do anything with it.

We can't merge the details into the original if it's not a subclass of `YetiException::Error`.  Well, it's Ruby, so we could open up the original exception instance and metaprogram the extra properties into it, but I don't think we want to do that.

Reply:
> I'm concerned that you'd need to be aware of it and poke deeper.

Perhaps we could add support to YetiLogger for walking the `.cause`s and logging them alongside the `YetiException::Error`. 

I share Matt's concern about obfuscating the original source of the error; it seems a little weird to have an error raised from a line other than what the stacktrace points us to. But with `cause` we can potentially have the best of both worlds: full access to the original exception _and_ the augmented data that `YetiException::Error` carries with it.